
<font size =6> Simple Exploratory Data Analysis using Python Pandas </font>
<font size=4> 
    
    
1. Importing datasets, dealing with missing values, changing data types.


2. Filtering, sorting, selecting specific column(s).


3. Dealing with duplicate values, dropping and adding rows and columns.


4. Counting values, counting unique values.</font>


<font size=5> Importing Pandas library & loading dataset</font>

In [82]:
# Importing pandas library
import pandas as pd

In [83]:
# loading dataset
df = pd.read_csv('datasets/train.csv')

In [84]:
# View first 10 rows of DataFrame using .head() method
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [87]:
# View last 10 rows of DataFrame using .tail() method
df.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

<font size=5> Finding the dimensionality of the dataset </font>

In [88]:
# Dimensionality of DataFrame - (Rows,Columns)
df.shape

(891, 12)

<font size=5> Filtering out the columns which are not useful for EDA from the dataset </font>

In [89]:
# usecols arguement to specify the column names that we want to work with
# index_col to specify colimn to use for index, default is 0
# .head() to display first 5 rows
df1 = pd.read_csv('datasets/train.csv', usecols=["PassengerId", "Survived", "Pclass", "Age", "Sex", "Embarked"], index_col=['PassengerId'])
df1.head()

Survived  Pclass     Sex   Age Embarked
PassengerId                                         
1                   0       3    male  22.0        S
2                   1       1  female  38.0        C
3                   1       3  female  26.0        S
4                   1       1  female  35.0        S
5                   0       3    male  35.0        S

Columns which are not related to find out survival probability. so these columns are dropped.

<font size=5> Displaying basic statistical details like percentile, mean, std etc. of a data frame or a series of numeric values</font>

In [90]:
# Stastical details of numeric values in dataset.
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

<font size=5>Printing information about a DataFrame including the index dtype and columns, non-null values and memory usage</font>

In [118]:
# Getting concise information about a DataFrame - index dtype and columns, non-null values & memory usage.
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 258 to 674
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null    int64   
 2   Pclass       891 non-null    int64   
 3   Name         891 non-null    object  
 4   Sex          891 non-null    object  
 5   Age          714 non-null    float64 
 6   SibSp        891 non-null    int64   
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float64 
 10  Cabin        204 non-null    object  
 11  Embarked     889 non-null    category
dtypes: category(1), float64(2), int64(5), object(4)
memory usage: 84.5+ KB




It indicates that there are total of 891 passesnger details among which 177 people's Age is missing and 687 people's Cabin details are missing.


<font size=5>Printing dtypes in the DataFrame</font>

In [92]:
# Getting dtypes in the DataFrame with .dtypes attribute.
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

<font size=5>Sorting the Data</font>

In [93]:
# Sorting the data by the values along either axis.
# Getting first 10 rows sorted according to the 'Fare' in non-decreasing order
df.sort_values('Fare').head(10)

PassengerId  Survived  Pclass                             Name   Sex  \
271          272         1       3     Tornquist, Mr. William Henry  male   
597          598         0       3              Johnson, Mr. Alfred  male   
302          303         0       3  Johnson, Mr. William Cahoone Jr  male   
633          634         0       1    Parr, Mr. William Henry Marsh  male   
277          278         0       2      Parkes, Mr. Francis "Frank"  male   
413          414         0       2   Cunningham, Mr. Alfred Fleming  male   
674          675         0       2       Watson, Mr. Ennis Hastings  male   
263          264         0       1            Harrison, Mr. William  male   
466          467         0       2            Campbell, Mr. William  male   
732          733         0       2             Knight, Mr. Robert J  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
271  25.0      0      0    LINE   0.0   NaN        S  
597  49.0      0      0    LINE   0.0   NaN        S  
302  19.0      0      0    LINE   0.0   NaN        S  
633   NaN      0      0  112052   0.0   NaN        S  
277   NaN      0      0  239853   0.0   NaN        S  
413   NaN      0      0  239853   0.0   NaN        S  
674   NaN      0      0  239856   0.0   NaN        S  
263  40.0      0      0  112059   0.0   B94        S  
466   NaN      0      0  239853   0.0   NaN        S  
732   NaN      0      0  239855   0.0   NaN        S

In [103]:
# Getting first 10 passengers who paid the highest ticket fares by using sort_values method with ascending = False attribute
# Here original data is not sorted
df.sort_values('Fare', ascending = False).head(10)


PassengerId  Survived  Pclass                                   Name  \
258          259         1       1                       Ward, Miss. Anna   
679          680         1       1     Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1                 Lesurer, Mr. Gustave J   
88            89         1       1             Fortune, Miss. Mabel Helen   
438          439         0       1                      Fortune, Mr. Mark   
27            28         0       1         Fortune, Mr. Charles Alexander   
341          342         1       1         Fortune, Miss. Alice Elizabeth   
742          743         1       1  Ryerson, Miss. Susan Parker "Suzette"   
311          312         1       1             Ryerson, Miss. Emily Borie   
118          119         0       1               Baxter, Mr. Quigg Edmond   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292              NaN        C  
679    male  36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292             B101        C  
88   female  23.0      3      2     19950  263.0000      C23 C25 C27        S  
438    male  64.0      1      4     19950  263.0000      C23 C25 C27        S  
27     male  19.0      3      2     19950  263.0000      C23 C25 C27        S  
341  female  24.0      3      2     19950  263.0000      C23 C25 C27        S  
742  female  21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
311  female  18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
118    male  24.0      0      1  PC 17558  247.5208          B58 B60        C

In [95]:
# Above sorting did not sorted the original dataset.
# Sorting DataFrame rows in descending order on "Fare" and using  inplace = True attribute
df.sort_values("Fare", ascending = False, inplace = True)
df.head()


PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
27            28         0       1      Fortune, Mr. Charles Alexander   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
88   female  23.0      3      2     19950  263.0000  C23 C25 C27        S  
27     male  19.0      3      2     19950  263.0000  C23 C25 C27        S

In [96]:
# Sorting Cabin column which has Missing (Nan) Values.
# By using na_postion = 'last' puts Nan at the end. default is 'last'
df.sort_values('Cabin', ascending = False, na_position = 'last').tail()


PassengerId  Survived  Pclass                             Name   Sex  \
633          634         0       1    Parr, Mr. William Henry Marsh  male   
413          414         0       2   Cunningham, Mr. Alfred Fleming  male   
822          823         0       1  Reuchlin, Jonkheer. John George  male   
732          733         0       2             Knight, Mr. Robert J  male   
674          675         0       2       Watson, Mr. Ennis Hastings  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
633   NaN      0      0  112052   0.0   NaN        S  
413   NaN      0      0  239853   0.0   NaN        S  
822  38.0      0      0   19972   0.0   NaN        S  
732   NaN      0      0  239855   0.0   NaN        S  
674   NaN      0      0  239856   0.0   NaN        S

<font size=5> Counting occurance in DataFrame</font>

In [97]:
# Finding the occurance of variable.
# Counting the number of male and female passengers with .value_counts()
df.Sex.value_counts()

male      577
female    314
Name: Sex, dtype: int64

<font size=5>Getting unique values in a DataFrame</font>

In [98]:
# Finding the unique values that occur in dataset or in a column
# Using .nunique() on dateset
df.nunique()


PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [99]:
# Using .nunique() on particular column 'Embarked'
df["Embarked"].nunique() 

3

<font size=5>After checking data types of column in data set by using 'df.dtypes', it is observed the Embarked column is object.
After counting the unique values in Embarked column with .unique(), we can see that there are 3 unique values in that column.
Hence it can be consider that its data type should be Categorical.</font>

In [100]:
# Changing Embarked column data tpe to Categorical by using .astype("category")
df['Embarked'] = df['Embarked'].astype('category')
df['Embarked'].dtype

CategoricalDtype(categories=['C', 'Q', 'S'], ordered=False)

In [106]:
# Filtering - finding passengers who Embarked from port 'Q'
df[df["Embarked"] == "Q"].head()

PassengerId  Survived  Pclass                         Name     Sex   Age  \
412          413         1       1       Minahan, Miss. Daisy E  female  33.0   
245          246         0       1  Minahan, Dr. William Edward    male  44.0   

     SibSp  Parch Ticket  Fare Cabin Embarked  
412      1      0  19928  90.0   C78        Q  
245      2      0  19928  90.0   C78        Q

In [109]:
# Filtering - Two or more conditions using AND, OR opertor
# Finding Female passengers having ticket fare < 100
df_fare_mask = df["Fare"] < 100
df_gender_mask = df["Sex"] == "female"
df[df_fare_mask & df_gender_mask].head()

PassengerId  Survived  Pclass  \
820          821         1       1   
520          521         1       1   
291          292         1       1   
486          487         1       1   
412          413         1       1   

                                                  Name     Sex   Age  SibSp  \
820  Hays, Mrs. Charles Melville (Clara Jennings Gr...  female  52.0      1   
520                              Perreault, Miss. Anne  female  30.0      0   
291            Bishop, Mrs. Dickinson H (Helen Walton)  female  19.0      1   
486    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  female  35.0      1   
412                             Minahan, Miss. Daisy E  female  33.0      1   

     Parch Ticket     Fare Cabin Embarked  
820      1  12749  93.5000   B69        S  
520      0  12749  93.5000   B73        S  
291      0  11967  91.0792   B49        C  
486      0  19943  90.0000   C93        S  
412      0  19928  90.0000   C78        Q

<font size=5> Checking if any value is Nan in a Dataframe</font>

In [120]:
# Finding Null Values using .isnull()
# Finding passengers whose is not known i.e Nan
null_mask = df["Cabin"].isnull() 
df[null_mask].head()

PassengerId  Survived  Pclass  \
258          259         1       1   
557          558         0       1   
380          381         1       1   
856          857         1       1   
708          709         1       1   

                                           Name     Sex   Age  SibSp  Parch  \
258                            Ward, Miss. Anna  female  35.0      0      0   
557                         Robbins, Mr. Victor    male   NaN      0      0   
380                       Bidois, Miss. Rosalie  female  42.0      0      0   
856  Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0      1      1   
708                        Cleaver, Miss. Alice  female  22.0      0      0   

       Ticket      Fare Cabin Embarked  
258  PC 17755  512.3292   NaN        C  
557  PC 17757  227.5250   NaN        C  
380  PC 17757  227.5250   NaN        C  
856     36928  164.8667   NaN        S  
708    113781  151.5500   NaN        S

In [112]:
# Finding the count of the passengers who cabin is not known using .isnull.sum() functions
df['Cabin'].isnull().sum()


687

<font size=5> As there are lot of *missing values* in the dataset columns. Hence, dropping 'Cabin' column which has 70% missing values.</font>


In [115]:
# Droping the 'Cabin' column 
df_without_cabin = df.drop(labels = ['Cabin'], axis=1)
df_without_cabin.head()

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
27            28         0       1      Fortune, Mr. Charles Alexander   

        Sex   Age  SibSp  Parch    Ticket      Fare Embarked  
258  female  35.0      0      0  PC 17755  512.3292        C  
737    male  35.0      0      0  PC 17755  512.3292        C  
679    male  36.0      0      1  PC 17755  512.3292        C  
88   female  23.0      3      2     19950  263.0000        S  
27     male  19.0      3      2     19950  263.0000        S